src/message_log.rs

```diff
@@ -131,131 +118,89 @@ impl PbftLog {
-            Err(PbftError::WrongNumMessages(
-                PbftMessageType::PrePrepare,
-                1,
-                pre_prep_msgs.len(),
-            ))
-        } else {
-            Ok(pre_prep_msgs)
-        }
-    }
+        let block_new_msg = block_new_msg.unwrap();
+        let preprep_msg = preprep_msg.unwrap();
 
-    fn check_log_prepare_msgs_match(
-        &self,
-        block_new_msgs: &[&ParsedMessage],
-        pre_prep_msgs: &[&ParsedMessage],
-        info: &PbftMessageInfo,
-    ) -> Result<(), PbftError> {
-        let prep_msgs = self.get_messages_of_type(
-            &PbftMessageType::Prepare,
-            info.get_seq_num(),
-            info.get_view(),
-        );
-        for prep_msg in prep_msgs {
-            // Make sure the contents match
-            if (!infos_match(prep_msg.info(), &pre_prep_msgs[0].info())
-                && prep_msg.get_block() != pre_prep_msgs[0].get_block())
-                || (!infos_match(prep_msg.info(), block_new_msgs[0].info())
-                    && prep_msg.get_block() != block_new_msgs[0].get_block())
-            {
-                return Err(PbftError::MessageMismatch(PbftMessageType::Prepare));
-            }
+        // Ensure BlockNew and PrePrepare match
+        if block_new_msg.get_block() != preprep_msg.get_block() {
+            error!(
+                "BlockNew {:?} does not match PrePrepare {:?}",
+                block_new_msg, preprep_msg
+            );
+            return Err(PbftError::MessageMismatch(PbftMessageType::PrePrepare));
         }
-        Ok(())
+
+        // Check if we have 2f + 1 matching Prepares
+        Ok(self.log_has_required_msgs(&PbftMessageType::Prepare, &preprep_msg, true, 2 * f + 1))
     }
 
     /// Checks if the node is ready to enter the `Committing` phase based on the `PbftMessage` received
     ///
     /// `check_committable` is true if for this node:
     ///   + `check_prepared` is true
-    ///   + This node has accepted `2f + 1` `Commit` messages, including its own
-    pub fn check_committable(&self, pbft_message: &ParsedMessage, f: u64) -> Result<(), PbftError> {
-        if check_msg_has_type(pbft_message, &PbftMessageType::Commit) {
-            return Err(PbftError::NotReadyForMessage);
+    ///   + This node has accepted `2f + 1` `Commit` messages, including its own, that match the
+    ///     corresponding `PrePrepare` message
+    pub fn check_committable(&self, info: &PbftMessageInfo, f: u64) -> Result<bool, PbftError> {
+        // Check if Prepared predicate is true
+        if !self.check_prepared(info, f)? {
+            return Ok(false);
         }
-        self.check_msg_against_log(&pbft_message, true, 2 * f + 1)?;
 
-        self.check_prepared(&pbft_message.as_msg_type(PbftMessageType::Prepare), f)?;
-        Ok(())
+        // Check if we have 2f + 1 matching Commits
+        let preprep_msg = self
+            .get_one_msg(info, &PbftMessageType::PrePrepare)
+            .unwrap();
```

Comment:
`get_one_msg` returns an `Option`, which you can't use a `?` for. And `unwrap` is safe here; we know this message exists because `check_preprepared` requires it.